components/net/http_cache.rs

```diff
@@ -0,0 +1,488 @@
+                        if max_age < age {
+                            return Duration::seconds(0i64);
+                        }
+                        return max_age - age;
+                    },
+                    _ => (),
+                }
+            }
+        }
+    }
+    if let Some(&header::Expires(header::HttpDate(t))) = response.headers.get::<header::Expires>() {
+        // store the period of time from now until expiry
+        let desired = t.to_timespec();
+        let current = time::now().to_timespec();
+        if desired > current {
+            return desired - current;
+        } else {
+            return Duration::seconds(0i64);
+        }
+    } else {
+       if let Some(_) = response.headers.get_raw("Expires") {
+           // Malformed Expires header, shouldn't be used to construct a valid response.
+           return Duration::seconds(0i64);
+       }
+    }
+    // Calculating Heuristic Freshness https://tools.ietf.org/html/rfc7234#section-4.2.2
+    if let Some((ref code, _)) = response.raw_status {
+        let heuristic_freshness = if let Some(&header::LastModified(header::HttpDate(t))) =
+            response.headers.get::<header::LastModified>() {
+            let last_modified = t.to_timespec();
+            let current = time::now().to_timespec();
+            (current - last_modified) / 10
+        } else {
+            // https://tools.ietf.org/html/rfc7234#section-5.5.4
+            // Since we do not generate such a warning, 24 hours is the max for heuristic calculation.
+            Duration::hours(24)
+        };
+        match *code {
+            200 | 203 | 204 | 206 | 300 | 301 | 404 | 405 | 410 | 414 | 501 => {
+                // Status codes that are cacheable by default https://tools.ietf.org/html/rfc7231#section-6.1
+                return heuristic_freshness
+            },
+            _ => {
+                // Other status codes can only use heuristic freshness if the public cache directive is present.
+                if let Some(&header::CacheControl(ref directives)) = response.headers.get::<header::CacheControl>() {
+                    let has_public_directive = directives.iter().any(|directive| {
+                        header::CacheDirective::Public == *directive
+                    });
+                    if has_public_directive {
+                        return heuristic_freshness;
+                    }
+                }
+            },
+        }
+    }
+    // Requires validation upon first use as default.
+    Duration::seconds(0i64)
+}
+
+/// Request Cache-Control Directives https://tools.ietf.org/html/rfc7234#section-5.2.1
+fn get_expiry_adjustment_from_request_headers(request: &Request, expires: Duration) -> Duration {
+    if let Some(directive_data) = request.headers.get_raw("cache-control") {
+        let directives_string = String::from_utf8(directive_data[0].to_vec()).unwrap();
+        let directives: Vec<&str> = directives_string.split(",").collect();
```

Comment:
Lots of uses of unwrap/collect/to_vec in this function that we should avoid if possible.

Reply:
Ok, I'll look into it...